### ISSUE TITLE
Incorrect trivial cast warning.

### ISSUE THREAD

**Issue by wdanilo** [rust-lang/rust-clippy#5128]
Hi, consider the following code:

```rust
#![warn(trivial_casts)]

pub fn test(t:String) {
    let _color = match &t as &str {
        "Toggle"  => "534666",
        "Gate"    => "e69d45",
        "Hold"    => "308695",
        "Lambda"  => "d45769",
        "Lambda2" => "d45769",
        _         => "455054",
    };
}
```

It warns:
```
warning: trivial cast: `&std::string::String` as `&str`
 --> src/lib.rs:4:24
  |
4 |     let _color = match &t as &str {
  |                        ^^^^^^^^^^
  |
note: lint level defined here
 --> src/lib.rs:1:9
  |
1 | #![warn(trivial_casts)]
  |         ^^^^^^^^^^^^^
  = help: cast can be replaced by coercion; this might require a temporary variable
```

The only way to refactor it and still make it compile is to:

```rust
#![warn(trivial_casts)]

pub fn test(t:String) {
    let variant = &t as &str;
    let _color = match variant {
        "Toggle"  => "534666",
        "Gate"    => "e69d45",
        "Hold"    => "308695",
        "Lambda"  => "d45769",
        "Lambda2" => "d45769",
        _         => "455054",
    };
}
```

However, this emits the warning as well. I haven't found any version of the code which does not require explicit `as &str`.

**Comment by lzutao** [comment]
You could write the snippet as:
```rust
#![warn(trivial_casts)]

pub fn test(t:String) {
    let _color = match t.as_str() {
        "Toggle"  => "534666",
        "Gate"    => "e69d45",
        "Hold"    => "308695",
        "Lambda"  => "d45769",
        "Lambda2" => "d45769",
        _         => "455054",
    };
}
```

Also, this is a lint defined by Rustc, not clippy.

**Comment by flip1995** [comment]
Yes, this is a rustc issue, please report it in the rust-lang/rust repo.
